lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,570 @@
+use lightning::ln::channelmanager::{ChannelDetails, PaymentSendFailure};
+use lightning::ln::features::InvoiceFeatures;
+use lightning::ln::msgs::LightningError;
+use lightning::routing::network_graph::NetworkGraph;
+use lightning::routing::router::{Route, RouteHint};
+use lightning::util::events::{Event, EventHandler};
+use lightning::util::logger::Logger;
+
+use std::cell::RefCell;
+use std::collections::hash_map::{self, HashMap};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+/// The maximum number of payment attempts before forwarding the [`Event::PaymentFailed`] to be
+/// handled.
+const MAX_PAYMENT_ATTEMPTS: usize = 3;
+
+/// A utility for paying [`Invoice]`s.
+pub struct InvoicePayer<P: Deref, R, G>
+where P::Target: Payer, R: Router, G: Deref<Target = NetworkGraph> {
+	payer: P,
+	router: R,
+	network_graph: G,
+	invoice_cache: Mutex<HashMap<PaymentHash, Invoice>>,
+}
+
+/// A trait defining behavior of an [`Invoice`] payer.
+pub trait Payer {
+	/// Returns the payer's node id.
+	fn node_id(&self) -> PublicKey;
+
+	/// Returns the payer's channels.
+	fn first_hops(&self) -> Vec<ChannelDetails>;
+
+	/// Sends a payment over the Lightning Network using the given [`Route`].
+	fn send_payment(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+	) -> Result<(), PaymentSendFailure>;
+}
+
+/// A trait defining behavior for routing an [`Invoice`] payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` using the provided first and last hops and
+	/// [`NetworkGraph`] for a payment with the given values.
+	fn find_route(
+		&self,
+		payer: &PublicKey,
+		network: &NetworkGraph,
+		payee: &PublicKey,
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+/// An [`EventHandler`] decorator for retrying failed payments.
+pub struct PaymentRetryHandler<P: Deref, R, G, L: Deref, E>
+where
+	P::Target: Payer,
+	R: Router,
+	G: Deref<Target = NetworkGraph>,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	invoice_payer: Arc<InvoicePayer<P, R, G>>,
+	payment_attempts: RefCell<HashMap<PaymentHash, usize>>,
+	logger: L,
+	event_handler: E,
+}
+
+impl<P: Deref, R, G> InvoicePayer<P, R, G>
+where P::Target: Payer, R: Router, G: Deref<Target = NetworkGraph> {
+	/// Creates a payer and a connected retry handler.
+	pub fn new<L: Deref, E: EventHandler>(
+		payer: P, router: R, network_graph: G, logger: L, event_handler: E
+	) -> (Arc<Self>, PaymentRetryHandler<P, R, G, L, E>)
```

Comment:
It occurs to me that it would be better to separate the construction of `InvoicePayer` and `PaymentRetryHandler`, as it wouldn't require using an `Arc` while still allowing it.